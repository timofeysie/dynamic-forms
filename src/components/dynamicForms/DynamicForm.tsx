import React, { useState } from 'react';
import './style.css';
import InputTextField from './inputs/InputTextField';
import DatepickerField from './inputs/DatepickerField';

const DynamicForm = (props: any) => {
  /* eslint-disable */
  // TODO: avoid the linting error:
  // "Must use destructuring props assignment react/destructuring-assignment"
  // fixing this causes the tests to break.
  const [ fields ] = useState(props.fields);
  //const [ input, setInput] = useState({});
  const [ input, setInput] = useState({});
  /* eslint-enable */

  const handleSubmit = (event: React.FormEvent<HTMLFormElement>): void => {
    event.preventDefault();
    props.onSubmit(input);
  };

  const handleChange = (event: React.FormEvent<HTMLInputElement>): void => {
    setInput({
      ...input,
      [event.currentTarget.name]: event.currentTarget.value,
    });
  };

  return (
    <form onSubmit={handleSubmit}>
      {fields.map((form: any) => {
        if (form.input_type === 'text') {
          return (
            <InputTextField
              validationString={form.validationString}
              validationMessage={form.validationMessage}
              name={form.name}
              placeholder={form.placeholder}
              required={form.required}
              key={form.placeholder}
              handleChange={(e: React.FormEvent<HTMLInputElement>) => {
                handleChange(e);
              }}
            />
          );
        }
        if (form.input_type === 'datepicker') {
          return (
            <DatepickerField
              name={form.name}
              required={form.required}
              key={form.placeholder+form.name}
              handleChange={(e: React.FormEvent<HTMLInputElement>) => {
                handleChange(e);
              }}
            />
          );
        }
        return <p>No form data.</p>;
      })}
      <input
        className="Submit"
        type="submit"
      />
    </form>
  );
};
export default DynamicForm;
