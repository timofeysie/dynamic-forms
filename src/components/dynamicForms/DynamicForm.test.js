import React from "react";
import { render, unmountComponentAtNode } from "react-dom";
import { act } from "react-dom/test-utils";
import ReactDOM from 'react-dom';
import DynamicForm from "./DynamicForm";
import { render as render2 } from "@testing-library/react";
import renderer from "react-test-renderer";

let container = null;
const fields = [
  {
    name: 'email',
    required: true,
    input_type: 'text',
    placeholder: 'email',
  },
];
beforeEach(() => {
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

it("renders a form with an input and submit button", () => {
  act(() => {
    render(<DynamicForm fields={fields}/>, container);
  });
  const form = container.querySelector('form');
  expect(form.length).toBe(2);

  const element1 = document.forms[0];
  expect(element1.children[0].tagName).toBe('DIV');
  expect(element1.children[0].children[0].type).toBe('text');
  expect(element1.children[1].tagName).toBe('INPUT');
  expect(element1.children[1].type).toBe('submit');
});

it("renders with react DOM", () => {
  const div = document.createElement("div");
  ReactDOM.render(<DynamicForm  fields={fields}/>, div);
});

// type="text"
it("renders an input with a type", () => {
  const { getByTestId } = render2(<DynamicForm fields={fields}/>);
  expect(getByTestId("inputid")).toHaveAttribute('type', 'text');
});

// name="email"
it("renders an input with a name equal to email", () => {
  const { getByTestId } = render2(<DynamicForm fields={fields}/>);
  expect(getByTestId("inputid")).toHaveAttribute('name', 'email');
});

// placeholder="email"
it("renders an input with a name equal to email", () => {
  const { getByTestId } = render2(<DynamicForm fields={fields}/>);
  expect(getByTestId("inputid")).toHaveAttribute('placeholder', 'email');
});

// required={true}
it("renders an input with a name equal to email", () => {
  const { getByTestId } = render2(<DynamicForm fields={fields}/>);
  expect(getByTestId("inputid")).toHaveAttribute('required');
});

// why are attribute and property the same???
it("renders an input with a type", () => {
  const { getByTestId } = render2(<DynamicForm fields={fields}/>);
  expect(getByTestId("inputid")).toHaveProperty('required');
});

it("matches snapshot", () => {
  const tree = renderer.create(<DynamicForm fields={fields}></DynamicForm>).toJSON();
  expect(tree).toMatchSnapshot();
})
