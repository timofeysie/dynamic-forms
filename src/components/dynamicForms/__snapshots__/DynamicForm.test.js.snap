// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<form
  onSubmit={[Function]}
>
  <div>
    <input
      autoCapitalize="off"
      data-testid="inputid"
      name="name"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="name"
      required={true}
      type="text"
    />
  </div>
  <div>
    <div
      className="react-datepicker-wrapper"
    >
      <div
        className="react-datepicker__input-container"
      >
        <input
          autoComplete="off"
          className=""
          disabled={false}
          name="dob"
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          readOnly={false}
          required={true}
          type="text"
          value="02/13/2020"
        />
      </div>
    </div>
  </div>
  <input
    className="Submit"
    type="submit"
  />
</form>
`;
